pkg/networkservice/common/discovercrossnse/server.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2021 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package discovercrossnse discovers forwarder from the registry.
+package discovercrossnse
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+	"github.com/networkservicemesh/sdk/pkg/tools/matchutils"
+)
+
+type discoverForwarderServer struct {
+	nseClient registry.NetworkServiceEndpointRegistryClient
+}
+
+// NewServer creates new instance of discovercrossnse networkservice.NetworkServiceServer.
+// Requires not nil nseClient.
+func NewServer(nseClient registry.NetworkServiceEndpointRegistryClient) networkservice.NetworkServiceServer {
+	if nseClient == nil {
+		panic("mseClient can not be nil")
+	}
+	return &discoverForwarderServer{
+		nseClient: nseClient,
+	}
+}
+
+func (d *discoverForwarderServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	var forwarderName = loadForwarderName(ctx)
+	var logger = log.FromContext(ctx).WithField("discoverForwarderServer", "request")
+
+	if forwarderName == "" {
+		stream, err := d.nseClient.Find(ctx, &registry.NetworkServiceEndpointQuery{
+			NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{
+				NetworkServiceNames: []string{
+					"forwarder",
+				},
+			},
+		})
+
+		if err != nil {
+			logger.Errorf("can not open registry nse stream by networkservice. Error: %v", err.Error())
+			return nil, errors.WithStack(err)
+		}
+
+		nses := registry.ReadNetworkServiceEndpointList(stream)
+
+		if len(nses) == 0 {
+			return nil, errors.New("forwarders not found")
+		}
+
+		nses = matchForwarders(request.Connection.GetLabels(), nses)
+
+		if len(nses) == 0 {
+			return nil, errors.New("no candidates found")
+		}
+
+		u, err := url.Parse(nses[0].Url) // TODO: Should we consider about load balancing?
```

Comment:
Yes, we should consider load balancing... open an issue :)

Reply:
I think this can be resolved in this PR if we go with https://github.com/networkservicemesh/sdk/pull/1122#discussion_r744229781

Reply:
How would it be resolved?

Reply:
1. cut this line https://github.com/networkservicemesh/sdk/blob/main/pkg/networkservice/common/roundrobin/server.go#L64 
2. add into` sdk/pkg/networkservice/setednpoint` chain element that will set this
https://github.com/networkservicemesh/sdk/blob/main/pkg/networkservice/common/roundrobin/server.go#L64 based on candidates and cleinturl from the context only if https://github.com/networkservicemesh/sdk/blob/main/pkg/networkservice/common/roundrobin/server.go#L64 is empty.
3.`discovercrossnse` should pass into context candates.

Finally, we can have chains like this:

nsmgr:
```go
...
discovercrossnse.NewServer(...) // works based on pass segments. If it is presented into connection just get it otherwise add canidates for the next load balancing chain element
roundrobind.NewServer()
```

forwarder

```go
discover.NewServer(...)
roundrobind.NewServer()
setednpoint.NewServer() //sets nse name into connection.
```

Reply:
Ah... I see... you are separating the setting of the value in the connection from the selection of where to go next.

Would it be possible to open an issue to come back to this and go ahead and land the P2MP now?

Reply:
> Would it be possible to open an issue to come back to this and go ahead and land the P2MP now?

Yes.

Reply:
Let's do that then :)